feat: added rnn layer and TimeSeries conversion

Gerhardsa0 opened the pull request:
Closes #614  
Closes #648
Closes #656 
Closes #601 
### Summary of Changes

<!-- Please provide a summary of changes in this pull request, ensuring all changes are explained. -->

- LSTM Layer
- conversion layer TimeSeries
- easier usage for conversion
- cleanup and refactoring, created TimeSeriesDataset
- docs were updated, as the TimeSeries no longer exist

Gerhardsa0 commented:
also right now the ```ClassificationNeurolNetwork``` takes a list of ```FNNLayer```
but this also in the doing by @sibre28

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -292,19 +295,23 @@ def __hash__(self) -> int:
         -------
-        size:
-            Size of this object in bytes.
+        Size of this object in bytes.
```
```suggestion
        size:
                Size of this object in bytes.
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -1271,3 +1278,54 @@ def plot_compare_time_series(self, time_series: list[TimeSeries]) -> Image:
+
+        Parameters
+        ----------
+        batch_size
```
```suggestion
        batch_size:
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -1271,3 +1278,54 @@ def plot_compare_time_series(self, time_series: list[TimeSeries]) -> Image:
+
+        Returns
+        -------
+        result :
```
```suggestion
        result:
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/ml/nn/_lstm_layer.py:
```diff
@@ -0,0 +1,67 @@
+
+        Returns
+        -------
+        result :
```
```suggestion
        result:
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/ml/nn/_lstm_layer.py:
```diff
@@ -0,0 +1,67 @@
+
+        Parameters
+        ----------
+        input_size
+            The number of neurons in the previous layer
+        output_size
```
```suggestion
        input_size:
            The number of neurons in the previous layer
        output_size:
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/ml/nn/_model.py:
```diff
@@ -58,6 +58,7 @@ def fit(
         if batch_size < 1:
             raise OutOfBoundsError(actual=batch_size, name="batch_size", lower_bound=ClosedBound(1))
         copied_model = copy.deepcopy(self)
+        # shouldnt the line below done by deepcopy?
```
Remove this comment?

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -292,19 +295,23 @@ def __hash__(self) -> int:
         Returns
         -------
-        hash:
+        hash : int
```
```suggestion
        hash:
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -292,19 +295,23 @@ def __hash__(self) -> int:
             The hash value.
         """
-        return _structural_hash(self.time, self.target, self.features, Table.__hash__(self))
+        return xxhash.xxh3_64(
+            hash(self.time).to_bytes(8)
+            + hash(self.target).to_bytes(8)
+            + hash(self.features).to_bytes(8)
+            + Table.__hash__(self).to_bytes(8),
+        ).intdigest()
```
Please use the new hashing method implemented in #609

Gerhardsa0 replied:
this was on old main version

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -1271,3 +1278,54 @@ def plot_compare_time_series(self, time_series: list[TimeSeries]) -> Image:
+        """
+        Return a Dataloader for the data stored in this time series, used for training neural networks.
+
+        It splits the target column into windows uses them as feature and creates targets for the time series, by
```
```suggestion
        It splits the target column into windows, uses them as feature and creates targets for the time series, by
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -1271,3 +1278,54 @@ def plot_compare_time_series(self, time_series: list[TimeSeries]) -> Image:
+        return DataLoader(dataset=_CustomDataset(np.array(x_s), np.array(y_s)), batch_size=batch_size)
+
+
+class _CustomDataset(Dataset):
+    def __init__(self, features: np.array, target: np.array):
+        self.X = torch.from_numpy(features.astype(np.float32))
+        self.Y = torch.from_numpy(target.astype(np.float32))
+        self.len = self.X.shape[0]
+
+    def __getitem__(self, item: int) -> tuple[torch.Tensor, torch.Tensor]:
+        return self.X[item], self.Y[item].unsqueeze(-1)
+
+    def __len__(self) -> int:
+        return self.len
```
Consider moving this Dataset and the Dataset of the TaggedTable into the Table class as both should be the same

Gerhardsa0 replied:
The TimeSeries class cant use the dataset from the Table class, because TimeSeries needs the Y vector.

lars-reimann commented on poetry.lock:
Please don't update dependencies in feature PRs.

lars-reimann commented on src/safeds/data/tabular/containers/_time_series.py:
```diff
@@ -5,10 +5,13 @@
 from typing import TYPE_CHECKING
 
 import matplotlib.pyplot as plt
+import numpy as np
 import pandas as pd
 import seaborn as sns
+import torch
+import xxhash
+from torch.utils.data import DataLoader, Dataset
```
In the `src` folder, imports of external libraries must now be moved into the functions that use them (see #624).

Gerhardsa0 commented:
@sibre28 can you check test_forward_workflow.py, if I used that correctly
 for some reason the neural network creates only zeros, we should check on this

Gerhardsa0 commented:
@lars-reimann 

I reworked the TimeSeries class due to serveral reasons:
- Compatibility with the Transformer

- I often used the as_table and time_table function for my use cases and I like the way that worked. Also this is now the workflow for TaggedTable so I adjusted it for TimeSeries aswell.

- TaggedTable and TimeSeries havbe similiar behavior.

There are some plotting functions which are still in this class, because they are mainly used by TimeSeries. Should this be moved somewhere else, or is the workflow still fine like that.
Also it reduced the code in LSTM-workflow and made it more easier to use.
I would like your feedback and thoughts, how it implemented right now.

lars-reimann commented:
@Gerhardsa0 Makes perfect sense this way (see also #693). We could move the remaining plot functions into the `Table` class:

* plot_lineplot: Exists already
* plot_scatterplot: Exists already
* plot_lagplot: Can be generalized to take two column names (1. temporal, 2. any)
* plot_compare_time_series: Can be generalized to compare any pair of columns in several tables

Gerhardsa0 commented:
> @Gerhardsa0 Makes perfect sense this way (see also #693). We could move the remaining plot functions into the `Table` class:
> 
>     * `plot_lineplot`: Exists already
Now removed form TimeSeries
> 
>     * `plot_scatterplot`: Exists already
Now removed form TimeSeries
> 
>     * `plot_lagplot`: Can be generalized to take two column names (1. temporal, 2. any)
added to table class, it takes one name and plots it against itself.
> 
>     * `plot_compare_time_series`: Can be generalized to compare any pair of numerical columns in several tables
added to Column class and compares multiple columns

lars-reimann commented:
> > ```
> > * `plot_lagplot`: Can be generalized to take two column names (1. temporal, 2. any)
> > ```
> 
> 
> added to table class, it takes one name and plots it against itself.

Should we instead add this to `Column` as well then? We currently don't have plot methods in `Table` that take one column name.

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+
+class TimeSeriesDataset:
+    """
+    A time series dataset maps feature and time columns to a target column. Not like the TableDataset a TimeSeries needs contain one target and one time column, but can have empty features.
```
```suggestion
    A time series dataset maps feature and time columns to a target column. Not like the TableDataset a TimeSeries needs to contain one target and one time column, but can have empty features.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+    ValueError
+        If the target column is also an extra column.
+    ValueError
+        If no feature columns remains.
```
```suggestion
        If no feature column remains.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        """
+        if not isinstance(other, TimeSeriesDataset):
+            return NotImplemented
+        return (
```
Small performance improvement:
```suggestion
        return (self is other) or (
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        return (
+            self.target == other.target
+            and self.features == other.features
+            and self._table == other._table
```
See #697 
```suggestion
            and self.extras == other.extras
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        hash:
+            The hash value.
+        """
+        return _structural_hash(self.target, self.features, self._table, self.time)
```
See #697 
```suggestion
        return _structural_hash(self.target, self.features, self.extras, self.time)
```

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        return (
+            sys.getsizeof(self._target)
+            + sys.getsizeof(self._features)
+            + sys.getsizeof(self._table)
```
See #697 
```suggestion
            + sys.getsizeof(self.extras)
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        ----------
+        window_size:
+            The size of the created windows
+
+        forecast_horizon:
+            The length of the forecast horizon, where all datapoints are collected until the given lag.
+
```
```suggestion
        forecast_horizon:
            The length of the forecast horizon, where all datapoints are collected until the given lag.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        batch_size:
+            The size of data batches that should be loaded at one time.
+
+
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        ----------
+        window_size:
+            The size of the created windows
+
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+
+        batch_size:
+            The size of data batches that should be loaded at one time.
+
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,323 @@
+        # every feature column wird auch gewindowed
+        # -> [i, win_size],[target]
+        feature_cols = self.features.to_columns()
+        for i in range(size - (forecast_horizon + window_size)):
+            window = target_np[i : i + window_size]
+            label = target_np[i + window_size + forecast_horizon]
+            for col in feature_cols:
+                data = col._data.to_numpy()
+                window = np.concatenate((window, data[i : i + window_size]))
+            x_s.append(window)
+            y_s.append(label)
```
Could you do this conversion directly in pytorch without the use of numpy? This could lead to faster conversion, especially when the user uses a gpu. The same question applies to all other conversions.

Gerhardsa0 replied:
yes, added

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_column.py:
```diff
@@ -1032,9 +1031,101 @@ def plot_histogram(self, *, number_of_bins: int = 10) -> Image:
+        ------
+        NonNumericColumnError
+            if the target column contains non numerical values
+
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_column.py:
```diff
@@ -1032,9 +1031,101 @@ def plot_histogram(self, *, number_of_bins: int = 10) -> Image:
+
+        ValueError
+            if the columns do not have the same size
+        >>> from safeds.data.tabular.containers import Column
```
```suggestion
        Examples
        --------
        >>> from safeds.data.tabular.containers import Column
```

Marsmaennchen221 commented on src/safeds/ml/classical/regression/_arima.py:
```diff
@@ -70,9 +71,10 @@ def fit(self, time_series: TimeSeries) -> ArimaModelRegressor:
-        if not isinstance(time_series, TimeSeries) and isinstance(time_series, Table):
-            raise NonTimeSeriesError
-        if time_series.number_of_rows == 0:
+        if not isinstance(time_series, TimeSeriesDataset) and isinstance(time_series, Table):
+            raise NonTimeSeriesDatasetError
```
Do you still need this, as the `TimeSeriesDataset` does no longer inherit the `Table`, so the value cannot be a `Table`

Gerhardsa0 replied:
this is correct, removed

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion.py:
```diff
@@ -29,6 +29,15 @@ def _data_conversion_fit(self, input_data: FT, batch_size: int, num_of_classes:
+    @abstractmethod
+    def _set_parameters(
+        self,
+        target_name: str,
+        time_name: str,
+        feature_names: list[str],
```
Consider using `**kwargs` as otherwise we have to add every parameter to this function and change this function every time we add a new, different input conversion

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,83 @@
+        Returns
+        -------
+        The size of the input for the neural network
+
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/ml/nn/_lstm_layer.py:
```diff
@@ -0,0 +1,90 @@
+        ValueError
+            If input_size < 1
+            If output_size < 1
+
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/ml/nn/_model.py:
```diff
@@ -102,16 +124,18 @@ def fit(
             for x, y in iter(dataloader):
                 optimizer.zero_grad()
-
+                # print("input:")
+                # print(str(x)+"\n")
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/ml/nn/_model.py:
```diff
@@ -102,16 +124,18 @@ def fit(
+                # print(str(x)+"\n")
                 pred = copied_model._model(x)
 
+                # print("output: "+str(pred)+ "\n")
+
                 loss = loss_fn(pred, y)
```
```suggestion
                loss = loss_fn(pred, y)
```

Marsmaennchen221 commented on src/safeds/ml/nn/_output_conversion_time_series.py:
```diff
@@ -0,0 +1,36 @@
+            The name of the new column where the prediction will be stored.
+        """
+        self._prediction_name = prediction_name
+        self._window_size = window_size
+        self._forecast_horizon = forecast_horizon
```
Are those always the same as in the `ÌnputConversionTimeSeries`? If so, you could automatically give the `OutputConversionTimeSeries` these parameters.

Gerhardsa0 replied:
Yes they are, adding it

Marsmaennchen221 commented on tests/safeds/ml/nn/test_lstm_workflow.py:
```diff
@@ -0,0 +1,46 @@
+    # Create a DataFrame
+    _inflation_path = "_datas/US_Inflation_rates.csv"
+    table = Table.from_csv_file(path=resolve_resource_path(_inflation_path))
+    # test_values = Table.from_rows(table.to_rows()[-165:])
```
```suggestion
```

Marsmaennchen221 commented on tests/safeds/ml/nn/test_lstm_workflow.py:
```diff
@@ -0,0 +1,46 @@
+    table = Table.from_csv_file(path=resolve_resource_path(_inflation_path))
+    # test_values = Table.from_rows(table.to_rows()[-165:])
+    rs = RangeScaler()
+    # ss_2 = RangeScaler()
+    # ss_2 = ss_2.fit(table, ["value"])
```
```suggestion
```

Marsmaennchen221 commented on tests/safeds/ml/nn/test_lstm_workflow.py:
```diff
@@ -0,0 +1,46 @@
+    # test_values.to_time_series_dataset("values", "date")
+    # suggest it ran through
+
+    # assert ts.plot_compare_time_series([test_values]) == snapshot_png_image
+    # assert ts.plot_lineplot() == snapshot_png_image
+    # assert test_values.plot_lineplot() == snapshot_png_image
```
Did you forget to re-enable these assertions?

Gerhardsa0 replied:
No these are not deterministic snapshots, even if I set a seed they fail from time to time. Due to some parameters may be set or not be set. So these snapshots may fail due to no reason.

Marsmaennchen221 commented on tests/safeds/ml/nn/test_lstm_workflow.py:
```diff
@@ -0,0 +1,46 @@
+    # assert ts.plot_compare_time_series([test_values]) == snapshot_png_image
+    # assert ts.plot_lineplot() == snapshot_png_image
+    # assert test_values.plot_lineplot() == snapshot_png_image
+    assert True
```
Could you define a better assertion?

Gerhardsa0 replied:
This just suggest, that it ran through

I have the code commented and note deleted, so it does not apply transformation all the time, but I still have the code there for future work.

Gerhardsa0 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,83 @@
+            batch_size,
+        )
+
+    def _set_parameters(
+        self,
+        target_name: str,
+        time_name: str,
+        feature_names: list[str],
+    ) -> None:
+        """Set the time_name variable for internal usage."""
+        self._time_name = time_name
+        self._feature_names = feature_names
+        self._target_name = target_name
```
Yes they are, adding it

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+
+class TimeSeriesDataset:
+    """
+    A time series dataset maps feature and time columns to a target column. Not like the TableDataset a TimeSeries needs to contain one target and one time column, but can have empty features.
```
```suggestion
    A time series dataset maps feature and time columns to a target column. Not like the TabularDataset a TimeSeries needs to contain one target and one time column, but can have empty features.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+    """
+    A time series dataset maps feature and time columns to a target column. Not like the TableDataset a TimeSeries needs to contain one target and one time column, but can have empty features.
+
+    Create a tabular dataset from a mapping of column names to their values.
```
```suggestion
    Create a time series dataset from a mapping of column names to their values.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+
+    def __eq__(self, other: object) -> bool:
+        """
+        Compare two tabular datasets.
```
```suggestion
        Compare two time series datasets.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+        Returns
+        -------
+        equals:
+            'True' if features and targets are equal, 'False' otherwise.
```
```suggestion
            'True' if features, time, target and extras are equal, 'False' otherwise.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+
+    def __hash__(self) -> int:
+        """
+        Return a deterministic hash value for this tabular dataset.
```
```suggestion
        Return a deterministic hash value for this time series dataset.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+
+    def to_table(self) -> Table:
+        """
+        Return a new `Table` containing the feature columns and the target column.
```
```suggestion
        Return a new `Table` containing the feature columns, the target column, the time column and the extra columns.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+        """
+        Return a new `Table` containing the feature columns and the target column.
+
+        The original `TabularDataset` is not modified.
```
```suggestion
        The original `TimeSeriesDataset` is not modified.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+        Returns
+        -------
+        table:
+            A table containing the feature columns and the target column.
```
```suggestion
            A table containing the feature columns, the target column, the time column and the extra columns.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+        Return a Dataloader for the data stored in this time series, used for training neural networks.
+
+        It splits the target column into windows, uses them as feature and creates targets for the time series, by
+        forecast length. The original table is not modified.
```
```suggestion
        forecast length. The original time series dataset is not modified.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+        for i in range(size - (forecast_horizon + window_size)):
+            window = target_tensor[i : i + window_size]
+            label = target_tensor[i + window_size + forecast_horizon]
+            print(window)
+            print(label)
```
```suggestion
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+        Return a Dataloader for the data stored in this time series, used for training neural networks.
+
+        It splits the target column into windows, uses them as feature and creates targets for the time series, by
+        forecast length. The original table is not modified.
```
```suggestion
        forecast length. The original time series dataset is not modified.
```

Marsmaennchen221 commented on src/safeds/data/labeled/containers/_time_series_dataset.py:
```diff
@@ -0,0 +1,330 @@
+
+    def _repr_html_(self) -> str:
+        """
+        Return an HTML representation of the tabular dataset.
```
```suggestion
        Return an HTML representation of the time series dataset.
```

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_column.py:
```diff
@@ -1032,9 +1031,102 @@ def plot_histogram(self, *, number_of_bins: int = 10) -> Image:
+        Returns
+        -------
+        plot:
+              A plot with all the Columns plotted by the ID on the x-axis.
```
```suggestion
            A plot with all the Columns plotted by the ID on the x-axis.
```

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_column.py:
```diff
@@ -1032,9 +1031,102 @@ def plot_histogram(self, *, number_of_bins: int = 10) -> Image:
+            if the target column contains non numerical values
+        ValueError
+            if the columns do not have the same size
+        Examples
```
```suggestion

        Examples
```

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_column.py:
```diff
@@ -1032,9 +1031,102 @@ def plot_histogram(self, *, number_of_bins: int = 10) -> Image:
+        Raises
+        ------
+        NonNumericColumnError
+            If the tcolumn contains non-numerical values.
```
```suggestion
            If the column contains non-numerical values.
```

Marsmaennchen221 commented on src/safeds/data/tabular/containers/_table.py:
```diff
@@ -2564,6 +2525,49 @@ def to_tabular_dataset(self, target_name: str, extra_names: list[str] | None = N
+        Returns
+        -------
+        dataset:
+            A new tabular dataset with the given target and feature names.
```
```suggestion
            A new time series dataset with the given target and feature names.
```

Marsmaennchen221 commented on src/safeds/exceptions/_ml.py:
```diff
@@ -95,13 +95,8 @@ def __init__(self) -> None:
-                " time and which are the target to predict.\n"
-            ),
-        )
+        super().__init__("Expected a instance of TimeSeriesDataset, got something else instead.")
```
```suggestion
        super().__init__("Expected an instance of TimeSeriesDataset, got something else instead.")
```

Marsmaennchen221 commented on src/safeds/ml/classical/regression/_arima.py:
```diff
@@ -70,9 +71,10 @@ def fit(self, time_series: TimeSeries) -> ArimaModelRegressor:
-        if not isinstance(time_series, TimeSeries) and isinstance(time_series, Table):
-            raise NonTimeSeriesError
-        if time_series.number_of_rows == 0:
+        if not isinstance(time_series, TimeSeriesDataset):
+            raise NonTimeSeriesDatasetError
```
Is this needed as the `time_series` value has to be of type `TimeSeriesDataset` as described in the signature?

Gerhardsa0 replied:
In theory the User could pass a other Dataset, because the signature does not forece it. But the ARIMA model is only useful on timeseries data

Gerhardsa0 replied:
removed, as signature is enough

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,80 @@
+
+        Returns
+        -------
+        The size of the input for the neural network
```
```suggestion
        size:
            The size of the input for the neural network
        
```

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,80 @@
+
+        Parameters
+        ----------
+        window_size
```
```suggestion
        window_size:
```

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,80 @@
+        ----------
+        window_size
+            The size of the created windows
+        forecast_horizon
```
```suggestion
        forecast_horizon:
```

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,80 @@
+        return self._is_fit_data_valid(input_data)
+
+    def _get_output_configuration(self) -> dict[str, Any]:
+        return {"window_size": self._window_size, "forecast_horizon": self._forecast_horizon}  # pragma: no cover
```
Why don't you test this in a test?

Gerhardsa0 replied:
it gets tested in the workflow, a test would the same

Gerhardsa0 replied:
added

Marsmaennchen221 commented on src/safeds/ml/nn/_lstm_layer.py:
```diff
@@ -0,0 +1,147 @@
+
+    def __eq__(self, other: object) -> bool:
+        """
+        Compare two convolutional transpose 2d layer.
```
```suggestion
        Compare two lstm layer.
```

Marsmaennchen221 commented on src/safeds/ml/nn/_lstm_layer.py:
```diff
@@ -0,0 +1,147 @@
+        Parameters
+        ----------
+        other:
+            The convolutional transpose 2d layer to compare to.
```
```suggestion
            The lstm layer to compare to.
```

Marsmaennchen221 commented on src/safeds/ml/nn/_lstm_layer.py:
```diff
@@ -0,0 +1,147 @@
+        Returns
+        -------
+        equals:
+            Whether the two convolutional transpose 2d layer are the same.
```
```suggestion
            Whether the two lstm layer are the same.
```

Marsmaennchen221 commented on src/safeds/ml/nn/_output_conversion_time_series.py:
```diff
@@ -0,0 +1,89 @@
+        Parameters
+        ----------
+        other:
+            The OutputConversionImage instance to compare to.
```
```suggestion
            The OutputConversionTimeSeries instance to compare to.
```

Marsmaennchen221 commented on src/safeds/ml/nn/_output_conversion_time_series.py:
```diff
@@ -0,0 +1,89 @@
+
+        Parameters
+        ----------
+        prediction_name
```
```suggestion
        prediction_name:
```

Marsmaennchen221 commented on src/safeds/ml/nn/_input_conversion_time_series.py:
```diff
@@ -0,0 +1,82 @@
+        return self._is_fit_data_valid(input_data)
+
+    def _get_output_configuration(self) -> dict[str, Any]:
+        return {"window_size": self._window_size, "forecast_horizon": self._forecast_horizon}  # pragma: no cover
```
```suggestion
        return {"window_size": self._window_size, "forecast_horizon": self._forecast_horizon}
```

lars-reimann commented:
:tada: This PR is included in version 0.24.0 :tada:

The release is available on:
- `v0.24.0`
- [GitHub release](https://github.com/Safe-DS/Library/releases/tag/v0.24.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
